File: ts/parser.ts
import type { IntermediateResult, ParseResult, Trie } from "./";
import type { IntHandler, StatHandlers } from "./types/handlers";
import type { Token } from "./types/stats";

export function tokenise(str: string) {
  return str.trim().toLocaleLowerCase("en-US").normalize().split(/\s+/);
}

export class Parser {
  constructor(
    readonly trie: Trie,
    readonly tokens: Record<string, Token[]>,
    readonly implied: Record<string, Record<string, number>>,
    readonly handlers: StatHandlers
  ) {
    for (const h of Object.values(handlers)) {
      if (!h.type) {
        console.log(h);
        throw new Error();
      }
    }
  }

  parse = (
    mod: string,
    log: (...args: unknown[]) => void = () => {}
  ): ParseResult[] => {
    const words = tokenise(mod);
    log(mod, words);
    const results: ParseResult[] = [];
    let index = 0;
    while (index < words.length) {
      const found = this.searchTrie(
        words[index],
        words,
        index + 1,
        this.trie,
        this.trie,
        log
      );
      if (!found.text || found.count === 0) {
        index = index + 1;
        continue;
      }
      const result: ParseResult = {
        text: found.text,
        stats: [],
      };
      results.push(result);
      this.processResult(found, result);

      index += found.count;
      log(found);
    }

    log(results);
    return results;
  };

  processResult = (input: IntermediateResult, output: ParseResult) => {
    this.processTokens(input, output);
    this.processImplications(input, output);
  };

  processTokens = (input: IntermediateResult, output: ParseResult) => {
    this.tokens[input.text!].forEach((t) => {
      if (t.type === "number") {
        const parsedValue = input.values.shift();
        if (parsedValue === undefined) {
          throw new Error("Missing value for stat" + t.stat);
        }
        const baseValue = (t.stat_value_handlers || []).reduceRight(
          (n, h) => this.reverseHandler(n, this.handlers[h] as IntHandler),
          parsedValue
        );
        output.stats.push({
          id: t.stat,
          index: t.index,
          parsedValue,
          baseValue,
        });
      } else if (t.type === "enum") {
        const parsedValue = input.values.shift();
        if (parsedValue === undefined) {
          throw new Error("Missing value for stat" + t.stat);
        }
        output.stats.push({
          id: t.stat,
          index: t.index,
          parsedValue,
          baseValue: parsedValue,
        });
      } else if (t.type === "nested") {
        const nested = input.nested;
        if (!nested?.text) {
          throw new Error("Missing nested stat for " + input.text);
        }
        output.text = output.text.replace("{0}", nested.text);
        this.processResult(nested, output);
      }
    });
  };

  processImplications(input: IntermediateResult, output: ParseResult) {
    const stats = this.implied[input.text!];
    if (!stats) return;
    for (const [id, value] of Object.entries(stats)) {
      if (value) {
        output.stats.push({
          id,
          index: -1,
          parsedValue: value,
          baseValue: value,
        });
      }
    }
  }

  reverseHandler(
    n: number,
    { addend = 0, multiplier = 1, divisor = 1 }: IntHandler
  ): number {
    return ((n - addend) * divisor) / multiplier;
  }

  searchTrie(
    word: string,
    words: string[],
    i: number,
    node: Trie,
    root: Trie,
    log: (...args: any[]) => void,
    canRecurse = true
  ): IntermediateResult {
    log(
      i,
      word,
      Object.keys(node.child_map || {}),
      Boolean(node.num_child),
      Boolean(node.any_child),
      node.terminal
    );

    if (i > words.length) {
      log(node);
      return {
        text: node.terminal,
        count: 0,
        values: node.stat_value ? [node.stat_value] : [],
      };
    }

    const results: IntermediateResult[] = [
      {
        text: node.terminal,
        count: 0,
        values: node.stat_value ? [node.stat_value] : [],
      },
    ];

    if (node.child_map && word in node?.child_map) {
      const result = this.searchTrie(
        words[i],
        words,
        i + 1,
        node.child_map[word],
        root,
        log
      );
      if (result.text) {
        if (node.stat_value) result.values.unshift(node.stat_value);
        result.count++;
        results.push(result);
      }
    }
    if (node.num_child) {
      const match = /^(([+-])?\d*\.?\d+)(\D.*)?$/.exec(word);
      log(match);
      if (match) {
        const [, num, prefix, suffix] = match;
        if (
          prefix &&
          node.child_map &&
          "+" in node.child_map &&
          node.child_map["+"].num_child
        ) {
          const result = this.searchTrie(
            word.substring(1),
            words,
            i,
            node.child_map["+"],
            root,
            log
          );
          if (result.text) results.push(result);
        }
        if (suffix) {
          const result = this.searchTrie(
            suffix,
            words,
            i,
            node.num_child,
            root,
            log
          );
          if (result.text) {
            result.values.unshift(parseFloat(num));
            results.push(result);
          }
        } else {
          const result = this.searchTrie(
            words[i],
            words,
            i + 1,
            node.num_child,
            root,
            log
          );
          if (result.text) {
            result.values.unshift(parseFloat(num));
            result.count++;
            results.push(result);
          }
        }
      }
    }

    if (node.stat_child && canRecurse) {
      const nested = this.searchTrie(word, words, i, root, root, log, false);
      log("nested stat", nested);
      if (nested.text) {
        const result = this.searchTrie(
          words[i + nested.count - 1],
          words,
          i + nested.count,
          node.stat_child,
          root,
          log
        );
        if (result.text) {
          result.count += nested.count;
          result.nested = nested;
          results.push(result);
        }
      }
    }
    if (node.any_child) {
      const result = this.searchTrie(
        words[i],
        words,
        i + 1,
        node.any_child,
        root,
        log
      );
      if (result) {
        result.count++;
        results.push(result);
      }
    }

    results.sort(
      (l, r) =>
        r.count - l.count ||
        r.values.length - l.values.length ||
        +!!r.text - +!!l.text ||
        r.text!.localeCompare(l.text!)
    );
    return results[0];
  }
}

File: rs/parser.rs
use poestat_static::{Trie, TRIE};
use regex::Regex;

use crate::result::IntermediateResult;

pub struct Parser<'w> {
    words: &'w Vec<String>,
    num_re: Regex,
}

impl<'w> Parser<'w> {
    pub fn new(words: &'w Vec<String>) -> Self {
        Parser {
            words,
            num_re: Regex::new(r"^(([+-])?\d*\.?\d+)(\D.*)?$").unwrap(),
        }
    }
    pub fn search(
        &self,
        word: Option<&String>,
        i: usize,
        node: &'static Trie,
        can_recurse: bool,
    ) -> IntermediateResult {
        #[cfg(test)]
        println!(
            "word: {:?}, i: {i}, children: {}, num: {}, stat: {}, any: {}",
            word,
            node.child_map.len(),
            node.num_child.is_some(),
            node.stat_child.is_some(),
            node.any_child.is_some()
        );
        if i > self.words.len() {
            return IntermediateResult {
                text: node.terminal,
                count: 0,
                values: node.stat_value.map_or_else(Vec::new, |v| vec![v as f32]),
                nested: None,
            };
        }

        let mut results = vec![];

        if let Some(v) = word.and_then(|w| node.child_map.get(w)) {
            let mut result = self.search(self.words.get(i), i + 1, v, can_recurse);
            if result.text.is_some() {
                if let Some(stat_value) = node.stat_value {
                    result.values.push(stat_value as f32);
                }
                result.count += 1;
                results.push(result);
            }
        }
        if let Some(num_child) = node.num_child {
            if let (Some(w), Some((prefix, num, suffix))) =
                (word, word.and_then(|w| self.match_number(w)))
            {
                if let Some(plus_node) = prefix.and_then(|_| node.child_map.get("+")) {
                    let result = self.search(Some(&w[1..].to_owned()), i, plus_node, can_recurse);
                    if result.text.is_some() {
                        results.push(result);
                    }
                }
                if let Some(Ok(num)) = num.map(|n| n.parse::<f32>()) {
                    if let Some(suffix) = suffix {
                        let mut result =
                            self.search(Some(&suffix.to_owned()), i, num_child, can_recurse);
                        if result.text.is_some() {
                            result.values.push(num);
                            results.push(result);
                        }
                    } else {
                        let mut result =
                            self.search(self.words.get(i), i + 1, num_child, can_recurse);
                        if result.text.is_some() {
                            result.values.push(num);
                            result.count += 1;
                            results.push(result);
                        }
                    }
                }
            }
        }

        if let Some(stat_child) = can_recurse.then_some(()).and(node.stat_child) {
            let nested = self.search(word, i, TRIE, false);
            if nested.text.is_some() {
                let j = i + nested.count;
                let mut result = self.search(self.words.get(j - 1), j, stat_child, can_recurse);
                if result.text.is_some() {
                    result.count += nested.count;
                    result.nested = Some(Box::new(nested));
                    results.push(result);
                }
            }
        }

        if let Some(any_child) = node.any_child {
            let mut result = self.search(self.words.get(i), i + 1, any_child, can_recurse);
            if result.text.is_some() {
                result.count += 1;
                results.push(result);
            }
        }

        results
            .into_iter()
            .max_by(|l, r| {
                l.count
                    .cmp(&r.count)
                    .then_with(|| l.values.len().cmp(&r.values.len()))
                    .then_with(|| l.text.cmp(&r.text))
            })
            .unwrap_or_else(|| IntermediateResult {
                text: node.terminal,
                count: 0,
                values: node.stat_value.map_or_else(Vec::new, |v| vec![v as f32]),
                nested: None,
            })
    }
    fn match_number<'v>(
        &self,
        val: &'v str,
    ) -> Option<(Option<&'v str>, Option<&'v str>, Option<&'v str>)> {
        self.num_re.captures(val).map(move |m| {
            (
                m.get(2).map(|m| m.as_str()),
                m.get(1).map(|m| m.as_str()),
                m.get(3).map(|m| m.as_str()),
            )
        })
    }
}

#[cfg(test)]
mod test {
    use super::*;
    #[test]
    fn test_num_re() {
        let v = vec![];
        let p = Parser::new(&v);
        assert_eq!(p.match_number("12"), Some((None, Some("12"), None)));
        assert_eq!(p.match_number("none"), None);
        assert_eq!(p.match_number("+12"), Some((Some("+"), Some("+12"), None)));
        assert_eq!(
            p.match_number("12suf"),
            Some((None, Some("12"), Some("suf")))
        );
    }

    #[test]
    fn test_parser() {
        let v = vec![
            "87%".to_string(),
            "increased".to_string(),
            "evasion".to_string(),
            "rating".to_string(),
        ];
        let p = Parser::new(&v);
        assert_eq!(
            p.search(Some(&v[0]), 1, TRIE, true),
            IntermediateResult {
                text: Some("{0}% increased Evasion Rating"),
                count: 4,
                values: vec![87.0],
                nested: None
            }
        );
    }

    #[test]
    fn test_recursive() {
        let v = [
            "added",
            "small",
            "passive",
            "skills",
            "grant:",
            "12%",
            "increased",
            "trap",
            "damage",
            "added",
            "small",
            "passive",
            "skills",
            "grant:",
            "12%",
            "increased",
            "mine",
            "damage",
        ]
        .iter()
        .map(|t| t.to_string())
        .collect();
        let p = Parser::new(&v);
        assert_eq!(
            p.search(Some(&v[0]), 1, TRIE, true),
            IntermediateResult {
                text: Some("Added Small Passive Skills grant: {0}"),
                count: 9,
                values: vec![],
                nested: Some(Box::new(IntermediateResult {
                    text: Some("{0}% increased Trap Damage"),
                    count: 4,
                    values: vec![12.0],
                    nested: None
                }))
            }
        );
    }

    #[test]
    fn test_enum() {
        let v = [
            "allocates",
            "gratuitous",
            "violence",
            "if",
            "you",
            "have",
            "the",
            "matching",
            "modifier",
            "on",
            "forbidden",
            "flame",
        ]
        .iter()
        .map(|t| t.to_string())
        .collect();
        let p = Parser::new(&v);
        assert_eq!(
            p.search(Some(&v[0]), 1, TRIE, true),
            IntermediateResult {
                text: Some("Allocates {0} if you have the matching modifier on Forbidden Flame"),
                count: 12,
                values: vec![27864.0],
                nested: None
            }
        );
    }
}
